app/scripts/controllers/metametrics.js

```diff
@@ -0,0 +1,415 @@
+    this.network = networkController.getNetworkIdentifier()
+    this.locale = prefState.currentLocale?.replace('_', '-') ?? ''
+    this.version =
+      environment === 'production' ? version : `${version}-${environment}`
+
+    this.store = new ObservableStore({
+      participateInMetaMetrics: null,
+      metaMetricsId: null,
+      metaMetricsSendCount: 0,
+      ...initState,
+    })
+
+    preferencesController.store.subscribe(({ currentLocale }) => {
+      this.locale = currentLocale?.replace('_', '-') ?? ''
+    })
+
+    networkController.on('networkDidChange', () => {
+      this.chainId = networkController.getCurrentChainId()
+      this.network = networkController.getNetworkIdentifier()
+    })
+
+    // We do not want to track events on development builds unless specifically
+    // provided a SEGMENT_WRITE_KEY. This also holds true for test environments and
+    // E2E, which is handled in the build process by never providing the SEGMENT_WRITE_KEY
+    // when process.env.IN_TEST is truthy
+    this.segment =
+      !segmentWriteKey || (isDevOrTestEnvironment && !segmentHost)
+        ? segmentMock ?? createSegmentMock(flushAt, flushInterval)
+        : new Analytics(segmentWriteKey, {
+            host: segmentHost,
+            flushAt,
+            flushInterval,
+          })
+
+    this.segmentLegacy =
+      !segmentLegacyWriteKey || (isDevOrTestEnvironment && !segmentHost)
+        ? segmentLegacyMock ?? createSegmentMock(flushAt, flushInterval)
+        : new Analytics(segmentLegacyWriteKey, {
+            host: segmentHost,
+            flushAt,
+            flushInterval,
+          })
+  }
+
+  generateMetaMetricsId() {
+    return bufferToHex(
+      sha3(
+        String(Date.now()) +
+          String(Math.round(Math.random() * Number.MAX_SAFE_INTEGER)),
+      ),
+    )
+  }
+
+  /**
+   * Setter for the `participateInMetaMetrics` property
+   *
+   * @param {boolean} bool - Whether or not the user wants to participate in MetaMetrics
+   * @returns {string|null} the string of the new metametrics id, or null if not set
+   *
+   */
+  setParticipateInMetaMetrics(bool) {
+    this.store.updateState({ participateInMetaMetrics: bool })
+    if (bool && !this.metaMetricsId) {
+      this.metaMetricsId = this.generateMetaMetricsId()
+    } else if (bool === false) {
+      this.metaMetricsId = null
+    }
+    return this.metaMetricsId
+  }
+
+  /**
+   * retrieve the user's preference for participating in MetaMetrics
+   */
+  get participateInMetaMetrics() {
+    return this.store.getState().participateInMetaMetrics
+  }
+
+  setMetaMetricsSendCount(val) {
+    this.store.updateState({ metaMetricsSendCount: val })
+  }
+
+  get metaMetricsSendCount() {
+    return this.store.getState().metaMetricsSendCount
+  }
+
+  get metaMetricsId() {
+    return this.store.getState().metaMetricsId
+  }
+
+  set metaMetricsId(metaMetricsId) {
+    this.store.updateState({ metaMetricsId })
```

Comment:
Oh! I was going to say, why is `metaMetricsId` being set as an instance field when it's in state? But it's not.

Do you find that using a setter like this is clearer than calling `updateState`? I found it to be a bit misleading.

Reply:
removed in favor of calling updateState directly in https://github.com/MetaMask/metamask-extension/pull/9857/commits/673a6508e7dc7c53cb04e6cd38829da33aae15d1